streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -468,55 +468,32 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
    * stream is uninterruptible.
    */
   final def catchAllCause[R1 <: R, E2, O1 >: O](f: Cause[E] => ZStream[R1, E2, O1]): ZStream[R1, E2, O1] = {
-    sealed abstract class State
-    object State {
-      case object NotStarted                         extends State
-      case class Self(finalizer: ZManaged.Finalizer) extends State
-      case object Other                              extends State
-    }
+    sealed abstract class State[+E]
+    case object NotStarted                                                   extends State[Nothing]
+    case class Self[E](finalizer: ZManaged.Finalizer, pull: Pull[R1, E, O1]) extends State[E]
+    case class Other(pull: Pull[R1, E2, O1])                                 extends State[Nothing]
 
     ZStream {
       for {
         finalizers <- ZManaged.releaseMap
-        selfPull   <- Ref.make[ZIO[R, Option[E], Chunk[O]]](Pull.end).toManaged_
-        otherPull  <- Ref.make[ZIO[R1, Option[E2], Chunk[O1]]](Pull.end).toManaged_
-        stateRef   <- Ref.make[State](State.NotStarted).toManaged_
+        ref        <- Ref.make[State[E]](NotStarted).toManaged_
         pull = {
-          def switch(releasePrev: ZManaged.Finalizer, e: Cause[Option[E]]): ZIO[R1, Option[E2], Chunk[O1]] = {
-            def next(e: Cause[E]) = ZIO.uninterruptible {
-              for {
-                _             <- releasePrev(Exit.fail(e))
-                tp            <- f(e).process.zio.provideSome[R1]((_, finalizers))
-                (release, as) = tp
-                _             <- otherPull.set(as)
-                _             <- stateRef.set(State.Other)
-                a             <- as
-              } yield a
-            }
-
-            Cause.sequenceCauseOption(e) match {
-              case None    => Pull.end
-              case Some(c) => next(c)
-            }
-          }
+          def use[R, E0, O](stream: ZStream[R, E0, O])(state: (ZManaged.Finalizer, Pull[R, E0, O]) => State[E]) =
+            stream.process.zio
+              .provideSome[R]((_, finalizers))
+              .flatMap { case (release, pull) => ref.set(state(release, pull)) *> pull }
+              .uninterruptible
```

Comment:
This was incorrectly marked as `uninterruptible` before. Good thing we're changing it :-)

Should be:
```suggestion
ZIO.uninterruptibleMask { restore =>
  restore(stream.process.zio)
    .provideSome[R]((_, finalizers))
    .flatMap { case (release, pull) => ref.set(state(release, pull)) *> restore(pull) }
}
```

Reply:
Done, thanks!